libafl_bolts/src/lib.rs

```diff
@@ -746,6 +746,31 @@ pub trait AsIterMut<'it> {
     fn as_iter_mut(&'it mut self) -> Self::IntoIter;
 }
 
+/// Create an `Iterator` that copies elements from a shared reference
+pub trait AsIterCopied<'it> {
+    /// The item type
+    type Item: 'it;
+    /// The iterator type
+    type IntoIter: Iterator<Item = Self::Item>;
+
+    /// Create an iterator from `&self`
+    fn as_copy_iter(&'it self) -> Self::IntoIter;
+}
+
+impl<'it, T> AsIterCopied<'it> for T
+where
+    T: AsIter<'it>,
+    <T as AsIter<'it>>::Item: Copy,
+{
+    type Item = <T as AsIter<'it>>::Item;
+
+    type IntoIter = core::iter::Copied<<T as AsIter<'it>>::IntoIter>;
+
+    fn as_copy_iter(&'it self) -> Self::IntoIter {
+        self.as_iter().copied()
+    }
+}
+
```

Comment:
I don't think we need these. Unless I'm misunderstanding the issue, it should be possible for you to define a type that merely creates the `Ref`/`RefMut` temporarily. e.g.,

```rs
struct MyRefCellMapIter<'a, T> {
  map: Ref<'a, MyMapRepr<T>>,
}
```

Reply:
Note that _not_ doing so removes the ability to use things like e.g. HitcountsMapObserver or the SIMD accel available if your map is accessible as a mutable slice after accessing via RefCell (which, btw, there is a wip that should enable this for your use case in #2061).

Reply:
Ahhh, I see.

Then, this strikes me as an issue with the definition of `AsIter` instead. I made the following changes:
```rs
/// Create an `Iterator` from a reference
pub trait AsIter<'it> {
    /// The item type
    type Item: 'it;
    /// The ref type
    type Ref: Deref<Target = Self::Item>;
    /// The iterator type
    type IntoIter: Iterator<Item = Self::Ref>;

    /// Create an iterator from &self
    fn as_iter(&'it self) -> Self::IntoIter;
}
// snip
impl<'a, 'it, T, const DIFFERENTIAL: bool> AsIter<'it> for StdMapObserver<'a, T, DIFFERENTIAL>
where
    T: Bounded
        + PartialEq
        + Default
        + Copy
        + Hash
        + 'static
        + Serialize
        + serde::de::DeserializeOwned
        + Debug,
{
    type Item = T;
    type Ref = &'it T;
    type IntoIter = Iter<'it, T>;

    fn as_iter(&'it self) -> Self::IntoIter {
        let cnt = self.usable_count();
        self.as_slice()[..cnt].iter()
    }
}
// snip
pub struct RefCellValueObserverIter<'it, T> {
    v: Ref<'it, [T]>,
}

impl<'it, T: Debug + Serialize + 'it> AsIter<'it> for RefCellValueObserver<'_, Vec<T>> {
    type Item = T;
    type Ref = Ref<'it, Self::Item>;
    type IntoIter = RefCellValueObserverIter<'it, T>;

    fn as_iter(&'it self) -> Self::IntoIter {
        Self::IntoIter {
            v: Ref::map(self.value.as_ref().borrow(), |v| v.as_slice()),
        }
    }
}

impl<'it, T: 'it> Iterator for RefCellValueObserverIter<'it, T> {
    type Item = Ref<'it, T>;

    fn next(&mut self) -> Option<Self::Item> {
        if self.v.is_empty() {
            return None;
        }
        let cloned = Ref::clone(&self.v);
        let (next, remainder) = Ref::map_split(cloned, |v| (&v[0], &v[1..]));
        self.v = remainder;
        Some(next)
    }
}
```

Apologies for the original confusion.

Reply:
(worth noting that `Copied` is actually only defined for &-references, but we can just `.map(|e| *e)` instead)

Reply:
This idea works perfectly for `AsIter`, but I still run into trouble when implementing `MapObserver::{get,set}`, because again, they require loaning out `&Self::Entry` with the same lifetime as `&self`. Do you think I should make `MapObserver` use a `Ref` and `RefMut` associated type as well?

Reply:
`get`/`set` should be replaced with copying versions, so no `&` for either of them is needed.

Reply:
Also, the SimpleIterator type should probably be deleted. If it's used anywhere, it should be replaced with `AsIter`

Reply:
Cherry-picking the `get`/`set` changes from this PR on top of the `AsIter`+`Ref` changes suggested above, I was able to implement `MapObserver` for `RefCellValueObserver<Vec<T>>`. Thanks for the feedback, @addisoncrump! I'm working now on getting approval to submit a PR with this new version.

Reply:
Awesome :slightly_smiling_face: To my knowledge, these changes should be fairly performant. Let me know how it goes and maybe we can make this the standard use case :chart_with_upwards_trend: